### PR TITLE
Correctly handle nil values in arrays

### PR DESCRIPTION
**Pull request by lafriks**
Was to late but posting it here as might be some other cases covered

### CODE REVIEW

File: repository_test.go [hunk 1]
```diff
@@ -2902,6 +2970,20 @@ func TestRepository_DeleteAll(t *testing.T) {
 	adapter.AssertExpectations(t)
 }
 
+func TestRepository_DeleteAllPointer(t *testing.T) {
+	var (
+		adapter = &testAdapter{}
+		repo    = New(adapter)
+		users   = []*User{{ID: 1}, nil}
```
Comment (Fs02) [inline repository_test.go:2977]: is this the edge case about nil check?

Comment (lafriks) [reply repository_test.go:2977]: yes

File: collection_test.go [hunk 2]
```diff
@@ -146,6 +146,24 @@ func TestCollection_Primary_usingElemInterface(t *testing.T) {
 	primariesCache.Delete(rt)
 }
 
+func TestCollection_PrimaryPointer_usingElemInterface(t *testing.T) {
+	var (
+		records = []*Item{
+			{UUID: "abc123"},
+			{UUID: "def456"},
+			nil,
```
Comment (lafriks) [inline collection_test.go:154]: also this

**Comment by lafriks** [comment]
Rebased code, please check `TestRepository_Update_ptrElem` test for regression now it panics (previously it did not)

File: document.go [hunk 3]
```diff
@@ -388,6 +398,9 @@ func newDocument(v interface{}, rv reflect.Value, readonly bool) *Document {
 		}
 		rt = rt.Elem()
 	}
+	if rt.Kind() == reflect.Ptr {
```
Comment (lafriks) [inline document.go:401]: This was removed in your PR, without this that test case panics

File: repository_test.go [hunk 4]
```diff
@@ -1307,6 +1307,39 @@ func TestRepository_Update(t *testing.T) {
...
+			"age":        Set("age", 0),
+			"created_at": Set("created_at", Now()),
+			"updated_at": Set("updated_at", Now()),
+		}
+		queries = From("users").Where(Eq("id", user.ID))
+	)
+
+	adapter.On("Update", queries, "id", mutates).Return(1, nil).Once()
+
+	assert.Nil(t, repo.Update(context.TODO(), &user))
```
Comment (Fs02) [inline repository_test.go:1332]: this case is double pointer? 🤔

File: document.go (hunk 3 above)
Comment (Fs02) [reply document.go:401]: what if the pointer contains pointer to nil?

File: repository_test.go (hunk 4 above)
Comment (Fs02) [reply repository_test.go:1332]: I'm not really sure if we should handle this case, since this case can be further extended to multiple level of pointers 🤔

File: document.go [hunk 5]
```diff
@@ -206,11 +211,16 @@ func (d Document) Value(field string) (interface{}, bool) {
 
 // SetValue of the field, it returns false if field does not exist, or it's not assignable.
 func (d Document) SetValue(field string, value interface{}) bool {
+	val := d.rv
```
Comment (Fs02) [inline document.go:214]: my idea was to only non pointer type inside document, so if document is created, d.rv always contains indirect value, not ptr

maybe we should fix the source that cause this issue? maybe I miss some place in collection? or is this related to nested pointer test?

File: document.go (hunk 3 above)
Comment (lafriks) [reply document.go]: wrong test case, sorry 🤦🏻‍♂️

File: repository_test.go (hunk 4 above)
Comment (lafriks) [reply repository_test.go]: yes, sorry, completely wrong test case 😅

File: document.go (hunk 5 above)
Comment (lafriks) [reply document.go]: removed, this was because of wrong test case

**Comment by lafriks** [comment]
Ok, removed everything except for handling nil values in arrays

File: collection.go [hunk 6]
```diff
@@ -73,11 +73,15 @@ func (c Collection) PrimaryValues() []interface{} {
...
+				values = make([]interface{}, 0, idxLen)
 			)
 
-			for j := range values {
-				values[j] = c.rvIndex(j).Field(index[i]).Interface()
+			for j := 0; j < idxLen; j++ {
+				if item := c.rv.Index(j); item.Kind() == reflect.Ptr && item.IsNil() {
+					continue
+				}
+				values = append(values, c.rvIndex(j).Field(index[i]).Interface())
```
Comment (Fs02) [inline collection.go:84]: to prevent double calls on rv.Index, how about something like this:
```suggestion
				if item := c.rvIndex(j); item.IsValid() {
					values = append(values, item.Field(index[i]).Interface())
				}
```

https://play.golang.org/p/Os_66nnUyUm

Comment (lafriks) [reply collection.go]: seems to do the job, fixed

**Review by Fs02 — APPROVED** [review]
LGTM 🚀 🚀
